### ISSUE TITLE
Predict call execution time degrades linearly with additional calls

### ISSUE THREAD

**Issue by NMRobert** [GPflow/GPflow#568]
- **OS Platform and Distribution**: Linux Ubuntu 17.04
- **TensorFlow installed from (source or binary)**: binary
- **Tensorflow version**: tensorflow-gpu==1.4.0
- **GPflow version:**  gpflow-1.0.0 (from github master source)
- **CUDNN installed:** 6
- **GPU:**  GeForce GTX 1080 Ti
- **Python Version:** Python 3.5.3

Issue:
Repeated calls to .predict (either .predict_f or .predict_y) have degrading execution time when called successively. The initial call might take perhaps 10ms, but later calls might take as long as several seconds each. This poses a problem when many calls to *.predict are desired. I did not observe this behaviour in versions of gpflow prior to 1.0.0.  This occurs for at least the regression models (GPR, SGPR, SVGP), and possibly others. 

```
import numpy as np
import matplotlib.pyplot as plt
import gpflow
import time

X = np.linspace(0,10, 100).reshape(-1,1)
y = np.sin(X) + np.random.normal(0, 0.25, size=100).reshape(-1,1)

k = gpflow.kernels.RBF(input_dim=1)
m = gpflow.models.gpr.GPR(X, y, kern=k)
m.compile()
gpflow.train.ScipyOptimizer().minimize(m)

print('Done optimize.')
predict_duration_log = []
for i in range(250):
    start = time.time()
    m.predict_y(X) 
    end = time.time()
    predict_duration_log.append(end-start)

plt.plot(predict_duration_log)
plt.show()
```
[
![image](https://user-images.githubusercontent.com/9734331/33043675-278a22a4-ce3e-11e7-84d0-7ed818b313a6.png)
](url)

**Comment by vdutor** [comment]
I replicated your experiment and came to the same conclusion. The code was running on CPU.

## GPflow 1.0
![gpflow_1_predict](https://user-images.githubusercontent.com/13780846/33067149-2d392af8-cea5-11e7-8213-8dcb910a30b0.png)

## GPflow 0.5
![gpflow_old_predict](https://user-images.githubusercontent.com/13780846/33067150-2d9d1626-cea5-11e7-9381-f32b00486570.png)

**Comment by alexggmatthews** [comment]
@NMRobert @vdutor thank you for reporting and confirming.
@awav I'm away from my desk but is this possibly because the graph is growing each time we call predict?

**Comment by awav** [comment]
That's very interesting, I hope it is not happening because of `report_uninitialized_variables`... In each run gpflow now check if variable needs re-initialization or not.

**Comment by awav** [comment]
Bad news, I was right...

**Comment by awav** [comment]
@alexggmatthews, @jameshensman and others :)

What I can do to fix it: turn off default initialization and checking by `tf.report_unitialized_variables`. But when you switch to another session, you will have to initialize the model explicitly. Any objections?

**Comment by awav** [comment]
I found a way to slow down speed degradation w/o removing auto-initializing feature:
<img width="573" alt="screen shot 2017-11-23 at 11 05 10" src="https://user-images.githubusercontent.com/24483645/33171218-54f2994a-d043-11e7-9ebe-f19a6f3ee91c.png">

**Comment by NMRobert** [comment]
@awav Thanks for your work on this! I'm not sure I fully understand the consequences of removing auto-initialisation; I don't personally think it would be an issue (for me), but might be confusing and/or unexpected behaviour for some users. What would the code for this look like?

**Comment by jameshensman** [comment]
I think that small hurdles for users who want to manage thier own sessions is okay. If you know enough to use a session, you know enough to initialize the tf variables.

**Comment by awav** [comment]
@NMRobert, When you switch from one session to another you will have to initialize parameters explicitly, for e.g.: `model.initialize(another_session, force=[True|False])`.

**Comment by markvdw** [comment]
@awav: The slower degradation, is that still present for "normal" users? I think this is quite seriously bad behaviour. For example, in my training code (which I let run for days), I run the prediction code periodically to monitor performance. It would be really annoying for this to slow down.

**Comment by awav** [comment]
I made short tests:

## **w/o** checking if variable were initialized (sec):

```python
In [10]: woi_pd.describe()
Out[10]:
count    50000.000000
mean         0.002592
std          0.003511
min          0.001846
25%          0.002330
50%          0.002432
75%          0.002665
max          0.388551
```

## **with** checking if variables were initialized (sec):

```python
In [11]: wi_pd.describe()
Out[11]:
count    50000.000000
mean         0.004250
std          0.003326
min          0.003169
25%          0.003954
50%          0.004069
75%          0.004368
max          0.405657
dtype: float64
```

Let's vote:
- *+1* leave auto-initializing as is
- *-1* we can live w/o initializing :)

**Comment by jameshensman** [comment]
+1

**Comment by markvdw** [comment]
Hi Artem, these benchmark times are really small. I'm only interested in the relative performance degradation for predictions which take around 5 or more seconds. If this will be the *absolute* degradation then I'm happy with the auto initialise. In my vote I have assumed this is the case, so if not, let me know :)

**Comment by awav** [comment]
@markvdw, it is not a degradation. I'm sorry for confusion - I edited an image above - my previous experiments were wrong. Here is new graphs:

<img width="551" alt="screen shot 2017-11-24 at 09 57 46" src="https://user-images.githubusercontent.com/24483645/33205228-128468ac-d0fe-11e7-9e28-b7dc45670ddc.png">

So, there is only small overhead 2-3ms, which is constant.

**Comment by markvdw** [comment]
Perfect, let's do auto init then! The only time when this would be a bad thing is if people call predict many times for small numbers of predictions. This isn't a good thing to do anyway, since we still recompute the Cholesky (annoyingly, there's still no good solution to this). I think there's no reason not to use auto-init.

Also, does this solution also fix the Iterator problem we spoke about?

**Comment by NMRobert** [comment]
I'm happy with the proposed solution that leaves auto-init intact.

**Comment by awav** [comment]
Here is results w/o spikes, just for the record:

```
In [44]: woi_pd.describe(percentiles=[0.75,0.9,0.999])
Out[44]:
count    49995.000000
mean         0.002557
std          0.000407
min          0.001846
50%          0.002432
75%          0.002665
90%          0.003011
99.9%        0.005564
max          0.013085
dtype: float64
```

```
In [46]: wi_pd.describe(percentiles=[0.75,0.9,0.999])
Out[46]:
count    49995.000000
mean         0.004218
std          0.000633
min          0.003169
50%          0.004068
75%          0.004368
90%          0.004831
99.9%        0.008685
max          0.042645
dtype: float64
```

<img width="567" alt="screen shot 2017-11-24 at 10 10 45" src="https://user-images.githubusercontent.com/24483645/33205841-439d053c-d100-11e7-9214-c53f887ef6a0.png">

**Comment by awav** [comment]
Resolved at #575
